include/analyzers/analyzer.h

```diff
@@ -48,71 +69,88 @@ class analyzer
 
     /**
      * Tokenizes a document.
-     * @param doc The document to store the tokenized information in
+     * @param doc The document to be tokenized
+     * @return a feature_map that maps the observed features to their
+     *  counts in the document
      */
-    virtual void tokenize(corpus::document& doc) = 0;
+    feature_map analyze(const corpus::document& doc);
 
     /**
      * Clones this analyzer.
      */
     virtual std::unique_ptr<analyzer> clone() const = 0;
 
-    /**
-     * @param config The config group used to create the analyzer from
-     * @return an analyzer as specified by a config object
-     */
-    static std::unique_ptr<analyzer> load(const cpptoml::table& config);
+    friend multi_analyzer<T>;
 
+  private:
     /**
-     * @param config The config group used to create the analyzer from
-     * @return the default filter chain for this version of MeTA,
-     * based on a config object
+     * The tokenization function that actually does the heavy lifting. This
+     * should be overridden in derived classes.
+     *
+     * @param doc The document to be tokenized
+     * @param counts The feature_map to place observed feature counts into
      */
-    static std::unique_ptr<token_stream>
-        default_filter_chain(const cpptoml::table& config);
+    virtual void tokenize(const corpus::document& doc, feature_map& counts) = 0;
+};
 
-    /**
-     * @param config The config group used to create the analyzer from
-     * @return the default filter chain for unigram words for this version
-     * of MeTA, based on a config object
-     */
-    static std::unique_ptr<token_stream>
-        default_unigram_chain(const cpptoml::table& config);
+/**
+ * Basic exception for analyzer interactions.
+ */
+class analyzer_exception : public std::runtime_error
+{
+  public:
+    using std::runtime_error::runtime_error;
+};
 
-    /**
-     * @param global The original config object with all parameters
-     * @param config The config group used to create the filters from
-     * @return a filter chain as specified by a config object
-     */
-    static std::unique_ptr<token_stream>
-        load_filters(const cpptoml::table& global,
-                     const cpptoml::table& config);
+/**
+ * @param config The config group used to create the analyzer from
+ * @return an analyzer as specified by a config object
+ */
+template <class T>
+std::unique_ptr<analyzer<T>> load(const cpptoml::table& config);
 
-    /**
-     * @param src The token stream that will feed into this filter
-     * @param config The config group used to create the filter from
-     * @return a single filter specified by a config object
-     */
-    static std::unique_ptr<token_stream>
-        load_filter(std::unique_ptr<token_stream> src,
-                    const cpptoml::table& config);
+extern template std::unique_ptr<analyzer<uint64_t>>
```

Comment:
I think the one-liner comments are good since this would seem a little unusual for someone that happens to see this.